channeldb/kvdb/etcd/db.go

```diff
@@ -34,6 +38,48 @@ func newEtcdBackend(host, user, pass string) (*db, error) {
 	return backend, nil
 }
 
+// View opens a database read transaction and executes the function f with the
+// transaction passed as a parameter.  After f exits, the transaction is rolled
+// back.  If f errors, its error is returned, not a rollback error (if any
+// occur).
+func (db *db) View(f func(tx kvdb.ReadTx) error) error {
+	apply := func(stm STM) error {
+		return f(newReadWriteTx(stm))
+	}
+
+	if err := RunSTM(db.cli, apply); err != nil {
+		// TODO: translate error?
+		return err
+	}
+	return nil
+
+}
+
+// Update opens a database read/write transaction and executes the function f
+// with the transaction passed as a parameter.  After f exits, if f did not
+// error, the transaction is committed.  Otherwise, if f did error, the
+// transaction is rolled back.  If the rollback fails, the original error
+// returned by f is still returned.  If the commit fails, the commit error is
+// returned.
+func (db *db) Update(f func(tx kvdb.RwTx) error) error {
+	apply := func(stm STM) error {
+		return f(newReadWriteTx(stm))
+	}
+	if err := RunSTM(db.cli, apply); err != nil {
+		// TODO: translate error?
+		return err
+	}
+	return nil
+}
+
+func (db *db) BeginReadWriteTx() (kvdb.RwTx, error) {
+	return newReadWriteTx(NewSTM(db.cli)), nil
+}
+
+func (db *db) BeginReadTx() (kvdb.ReadTx, error) {
+	return newReadWriteTx(NewSTM(db.cli)), nil
```

Comment:
are we able to assert that the db isn't modified?

Reply:
Not really, what we can do is call `Commit` then retry manually. In general `BeginReadTx` and `BeginReadWriteTx` isn't really nice as we need to handle errors which the current interface doesn't support in all cases, whereas we can do this in `Update` and `View`